src/Symfony/Component/Scheduler/Worker.php

```diff
@@ -0,0 +1,177 @@
+
+    /**
+     * @var LoggerInterface|null
+     */
+    private $logger;
+
+    /**
+     * @var PersistingStoreInterface|BlockingStoreInterface|null
+     */
+    private $store;
+
+    /**
+     * @param BlockingStoreInterface|PersistingStoreInterface|null $store
+     */
+    public function __construct(array $runners, TaskExecutionTracker $tracker, EventDispatcherInterface $eventDispatcher = null, LoggerInterface $logger = null, $store = null)
+    {
+        $this->runners = $runners;
+        $this->tracker = $tracker;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->logger = $logger;
+        $this->store = $store;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function execute(TaskInterface $task): void
+    {
+        $this->dispatchEvent(new WorkerStartedEvent($this));
+        $this->checkTaskState($task);
+
+        foreach ($this->runners as $runner) {
+            if (!$runner->support($task)) {
+                continue;
+            }
+
+            $this->handleChainedTask($task);
+
+            $lockedTask = $this->lockTask($task);
+
+            try {
+                if ($lockedTask->acquire()) {
+                    $this->tracker->watch($task);
+                    $this->handleTask($runner, $task);
+                    $this->tracker->endWatch($task);
+                    $lockedTask->release();
+
+                    return;
+                }
+            } catch (\Error $error) {
+                $this->failedTasks = new TaskList([new FailedTask($task, $error->getMessage())]);
+            }
+        }
+
+        throw new UndefinedRunnerException(sprintf('No runner found for the given task "%s"', $task->getName()));
+    }
+
+    private function checkTaskState(TaskInterface $task): void
+    {
+        if (State::UNSPECIFIED === $task->get('state')) {
+            throw new \LogicException('The task state must be defined in order to be executed!');
+        }
+
+        if (\in_array($task->get('state'), [State::PAUSED, State::DISABLED])) {
+            $this->log(sprintf('The following task "%s" is paused|disabled, consider enable it if it should be executed!', $task->getName()));
+
+            return;
+        }
+    }
+
+    private function dispatchEvent(Event $event): void
+    {
+        if (null === $this->eventDispatcher) {
+            return;
+        }
+
+        $this->eventDispatcher->dispatch($event);
+    }
+
+    private function lockTask(TaskInterface $task): LockInterface
+    {
+        if (null === $this->store) {
+            $this->store = new FlockStore();
+        }
+
+        $factory = new LockFactory($this->store);
+
+        return $factory->createLock($task->getName());
+    }
+
+    private function handleTask(RunnerInterface $runner, TaskInterface $task): void
+    {
+        $this->dispatchEvent(new TaskToExecuteEvent($task));
+        $output = $runner->run($task);
+        $task->set('last_execution', new \DateTimeImmutable());
```

Comment:
shouldnt this be inside an event subscriber?

Reply:
Depends on the previous comment